syncer/syncer.go

```diff
@@ -0,0 +1,402 @@
+		var layer *types.Layer
+		var err error
+		if layer, err = s.syncLayer(ctx, layerID); err != nil {
+			logger.With().Error("failed to sync to layer", log.Err(err))
+			return false
+		}
+
+		if len(layer.Blocks()) == 0 {
+			logger.Info("setting layer %v to zero-block", layerID)
+			if err := s.mesh.SetZeroBlockLayer(layerID); err != nil {
+				logger.With().Error("failed to set zero-block for layer", layerID, log.Err(err))
+				return false
+			}
+		}
+
+		if s.shouldValidateLayer(layerID) {
+			vQueue <- layer
+		}
+		logger.Info("finished data sync for layer %v", layerID)
+	}
+	s.setStateAfterSync(ctx)
+	logger.With().Info("node is synced. waiting for validation",
+		log.FieldNamed("current", s.ticker.GetCurrentLayer()),
+		log.FieldNamed("latest", s.mesh.LatestLayer()),
+		log.FieldNamed("validated", s.mesh.ProcessedLayer()))
+	return true
+}
+
+func (s *Syncer) setStateBeforeSync(ctx context.Context) {
+	current := s.ticker.GetCurrentLayer()
+	if current == 0 {
+		s.setSyncState(ctx, synced)
+	}
+	latest := s.mesh.LatestLayer()
+	if current > latest && current-latest >= outOfSyncThreshold {
+		s.logger.WithContext(ctx).With().Info("node is too far behind",
+			log.FieldNamed("current", current),
+			log.FieldNamed("latest", latest),
+			log.FieldNamed("behindThreshold", outOfSyncThreshold))
+		s.setSyncState(ctx, notSynced)
+	}
+}
+
+func (s *Syncer) setStateAfterSync(ctx context.Context) {
+	currSyncState := s.getSyncState()
+	current := s.ticker.GetCurrentLayer()
+	// if we have gossip-synced to the target synced layer, we are ready to participate in the consensus
+	if currSyncState == gossipSync && s.getTargetSyncedLayer() <= current {
+		s.setSyncState(ctx, synced)
+	} else if currSyncState == notSynced {
+		// wait till s.ticker.GetCurrentLayer() + numGossipSyncLayers to participate in consensus
+		s.setSyncState(ctx, gossipSync)
+		s.setTargetSyncedLayer(ctx, current+numGossipSyncLayers)
+	}
+}
+
+func (s *Syncer) syncLayer(ctx context.Context, layerID types.LayerID) (*types.Layer, error) {
+	if s.isClosed() {
+		return nil, errors.New("shutdown")
+	}
+
+	layer, err := s.getLayerFromPeers(ctx, layerID)
+	if err != nil {
+		s.logger.WithContext(ctx).With().Error("failed to get layer from neighbors", layerID, log.Err(err))
+		return nil, err
+	}
+
+	s.getATXs(ctx, layerID)
+	return layer, nil
+}
+
+func (s *Syncer) getLayerFromPeers(ctx context.Context, layerID types.LayerID) (*types.Layer, error) {
+	ch := s.fetcher.PollLayer(ctx, layerID)
+	res := <-ch
+	if res.Err != nil {
+		if res.Err == layerfetcher.ErrZeroLayer {
+			return types.NewLayer(layerID), nil
+		}
+		return nil, res.Err
+	}
+	return s.mesh.GetLayer(layerID)
+}
+
+func (s *Syncer) getATXs(ctx context.Context, layerID types.LayerID) {
+	if layerID.GetEpoch() == 0 {
+		s.logger.WithContext(ctx).Info("skip getting ATX in epoch 0")
+		return
+	}
+	lastLayerOfEpoch := (layerID.GetEpoch() + 1).FirstLayer() - 1
+	if layerID != lastLayerOfEpoch {
```

Comment:
we need to sync ATX also for the current epoch
consider that we might become synced in the middle of an epoch so this condition will not met but still - all ATXs published in this epoch must be synced.

Reply:
done. 

note that this is different behavior from the old syncer impl. but `dockertest-sync-elk` passed on this change.